### PR TITLE
Fixing removal from gas price when moving future->current

### LINKED ISSUE #2020: Inconsistency in transaction queue
Pretty easily reproducible on Morden (in debug mode; perhaps also on mainnet)
```
thread 'IO Worker #2' panicked at 'assertion failed: `(left == right)` (left: `2`, right: `1`)', ethcore/src/miner/transaction_queue.rs:253
stack backtrace:
   1:     0x562c6b2fda33 - std::sys::backtrace::tracing::imp::write::h482d45d91246faa2
   2:     0x562c6b30545d - std::panicking::default_hook::_{{closure}}::h89158f66286b674e
   3:     0x562c6b304320 - std::panicking::default_hook::h9e30d428ee3b0c43
   4:     0x562c6b304928 - std::panicking::rust_panic_with_hook::h2224f33fb7bf2f4c
   5:     0x562c6b3047c2 - std::panicking::begin_panic::hcb11a4dc6d779ae5
   6:     0x562c6b304700 - std::panicking::begin_panic_fmt::h310416c62f3935b3
   7:     0x562c6a6d8001 - ethcore::miner::transaction_queue::TransactionSet::insert::h2b4f89990d548453
                        at /home/tomusdrw/eth/parity/<std macros>:8
   8:     0x562c6a6dd4f3 - ethcore::miner::transaction_queue::TransactionQueue::move_all_to_future::hdd56fd1ba97f3c3d
                        at /home/tomusdrw/eth/parity/ethcore/src/miner/transaction_queue.rs:649
   9:     0x562c6a6dbcbd - ethcore::miner::transaction_queue::TransactionQueue::remove_all::h280ba9e1aeb8004a
                        at /home/tomusdrw/eth/parity/ethcore/src/miner/transaction_queue.rs:569
  10:     0x562c6a7f2ead - _<ethcore..miner..miner..Miner as ethcore..miner..MinerService>::chain_new_blocks::_{{closure}}::he123ed11bf40cc7b
                        at /home/tomusdrw/eth/parity/ethcore/src/miner/miner.rs:892
  11:     0x562c6a0e7458 - _<rayon..par_iter..for_each..ForEachConsumer<'f, OP> as rayon..par_iter..internal..Folder<ITEM>>::consume::h0d6fa459ba0bc57f
                        at /home/tomusdrw/.multirust/toolchains/nightly/cargo/registry/src/github.com-1ecc6299db9ec823/rayon-0.3.1/src/par_iter/for_each.rs:45
  12:     0x562c6a0ec351 - _<rayon..par_iter..map..MapFolder<'m, C, MAP_OP> as rayon..par_iter..internal..Folder<ITEM>>::consume::h9f3b46da382e2b88
                        at /home/tomusdrw/.multirust/toolchains/nightly/cargo/registry/src/github.com-1ecc6299db9ec823/rayon-0.3.1/src/par_iter/map.rs:201
  13:     0x562c6a4862b9 - rayon::par_iter::internal::bridge_producer_consumer::h9c471bc5d80ea683
                        at /home/tomusdrw/.multirust/toolchains/nightly/cargo/registry/src/github.com-1ecc6299db9ec823/rayon-0.3.1/src/par_iter/internal.rs:120
  14:     0x562c6a0d1507 - _<rayon..par_iter..internal..bridge..Callback<C> as rayon..par_iter..internal..ProducerCallback<ITEM>>::callback::h7cb7597affe7b6e1
                        at /home/tomusdrw/.multirust/toolchains/nightly/cargo/registry/src/github.com-1ecc6299db9ec823/rayon-0.3.1/src/par_iter/internal.rs:95
  15:     0x562c6a0cbfce - _<rayon..par_iter..slice..SliceIter<'data, T> as rayon..par_iter..IndexedParallelIterator>::with_producer::h300f61133f8dc678
                        at /home/tomusdrw/.multirust/toolchains/nightly/cargo/registry/src/github.com-1ecc6299db9ec823/rayon-0.3.1/src/par_iter/slice.rs:67
  16:     0x562c6a48656a - rayon::par_iter::internal::bridge::h578a99159d4005c5
                        at /home/tomusdrw/.multirust/toolchains/nightly/cargo/registry/src/github.com-1ecc6299db9ec823/rayon-0.3.1/src/par_iter/internal.rs:78
  17:     0x562c6a03f265 - _<rayon..par_iter..slice..SliceIter<'data, T> as rayon..par_iter..ParallelIterator>::drive_unindexed::h78be15417c85e235
                        at /home/tomusdrw/.multirust/toolchains/nightly/cargo/registry/src/github.com-1ecc6299db9ec823/rayon-0.3.1/src/par_iter/slice.rs:41
  18:     0x562c6a02d0c5 - _<rayon..par_iter..map..Map<M, MAP_OP> as rayon..par_iter..ParallelIterator>::drive_unindexed::h6f2e373929b8ef2a
                        at /home/tomusdrw/.multirust/toolchains/nightly/cargo/registry/src/github.com-1ecc6299db9ec823/rayon-0.3.1/src/par_iter/map.rs:28
  19:     0x562c6a485629 - rayon::par_iter::for_each::for_each::h59712da6da618c3b
                        at /home/tomusdrw/.multirust/toolchains/nightly/cargo/registry/src/github.com-1ecc6299db9ec823/rayon-0.3.1/src/par_iter/for_each.rs:11
  20:     0x562c6a48545d - rayon::par_iter::ParallelIterator::for_each::h511d6c2b62ed46c4
                        at /home/tomusdrw/.multirust/toolchains/nightly/cargo/registry/src/github.com-1ecc6299db9ec823/rayon-0.3.1/src/par_iter/mod.rs:101
  21:     0x562c6a6d6184 - _<ethcore..miner..miner..Miner as ethcore..miner..MinerService>::chain_new_blocks::h81d9e094245a155a
                        at /home/tomusdrw/eth/parity/ethcore/src/miner/miner.rs:883
  22:     0x562c6a674a2d - ethcore::client::client::Client::import_verified_blocks::h72fd5a938020d96e
                        at /home/tomusdrw/eth/parity/ethcore/src/client/client.rs:394
  23:     0x562c6a6b69e3 - _<ethcore..service..ClientIoHandler as ethcore_io..IoHandler<ethcore..service..ClientIoMessage>>::message::hdd1f46e03d0e9dac
                        at /home/tomusdrw/eth/parity/ethcore/src/service.rs:164
  24:     0x562c6a036fbc - ethcore_io::worker::Worker::do_work::hcd9fc73c319adec0
                        at /home/tomusdrw/eth/parity/util/io/src/worker.rs:117
  25:     0x562c6a037530 - ethcore_io::worker::Worker::work_loop::h0c83488e1b1a5b49
                        at /home/tomusdrw/eth/parity/util/io/src/worker.rs:97
  26:     0x562c6a7eb821 - ethcore_io::worker::Worker::new::_{{closure}}::_{{closure}}::h4719b95c4dbfe5ce
                        at /home/tomusdrw/eth/parity/util/io/src/worker.rs:72
  27:     0x562c6a036012 - ethcore_io::panics::PanicHandler::catch_panic::h7dd7b481458278a3
                        at /home/tomusdrw/eth/parity/util/io/src/panics.rs:85
  28:     0x562c6a7ebc91 - ethcore_io::worker::Worker::new::_{{closure}}::hc49a19488a28719e
                        at /home/tomusdrw/eth/parity/util/io/src/worker.rs:71
  29:     0x562c6a5f8620 - _<std..panic..AssertUnwindSafe<F> as core..ops..FnOnce<()>>::call_once::h9b39a1897fe740f1
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/libstd/panic.rs:256
  30:     0x562c6a2677e3 - std::panicking::try::do_call::hcf7ca47344f6173d
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/libstd/panicking.rs:327
  31:     0x562c6b30cf26 - __rust_maybe_catch_panic
  32:     0x562c6a266089 - std::panicking::try::hab203785b94f9d0d
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/libstd/panicking.rs:303
  33:     0x562c6a25cb84 - std::panic::catch_unwind::he83b420176ee0945
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/libstd/panic.rs:312
  34:     0x562c6a812b08 - std::thread::Builder::spawn::_{{closure}}::h41ec89490d4b16aa
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/libstd/thread/mod.rs:284
  35:     0x562c6a4268ae - _<F as alloc..boxed..FnBox<A>>::call_box::h0856070e6ba7ae20
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/liballoc/boxed.rs:591
  36:     0x562c6b3030f0 - std::sys::thread::Thread::new::thread_start::he0bf102845911132
  37:     0x7fa302fd7183 - start_thread
  38:     0x7fa303a0137c - clone
  39:                0x0 - <unknown>
```

### PR DESCRIPTION
**Pull request by tomusdrw**
Extracted `GasPriceQueue` operations into a separate struct.

Closes #2020

### CODE REVIEW

File: ethcore/src/miner/transaction_queue.rs [hunk 1]
```diff
@@ -219,6 +220,47 @@ impl VerifiedTransaction {
...
+	/// Remove an item from a BTreeMap/HashSet "multimap".
+	/// Returns true if the item was removed successfully.
+	pub fn remove(&mut self, gas_price: &U256, hash: &H256) -> bool {
+		if let Some(mut hashes) = self.backing.get_mut(gas_price) {
+			let only_one_left = hashes.len() == 1;
+			if !only_one_left {
+				// Operation may be ok: only if hash is in gas-price's Set.
+				return hashes.remove(hash);
+			}
+			if hashes.iter().next().unwrap() != hash {
```
Comment (rphmeier) [inline ethcore/src/miner/transaction_queue.rs]: should be an `expect` here with a proof

Comment (keorn) [inline ethcore/src/miner/transaction_queue.rs]: `.expect("Look at the 4 lines above")`

Comment (rphmeier) [inline ethcore/src/miner/transaction_queue.rs]: it's a trivial proof but it's important to be consistent

Comment (tomusdrw) [inline ethcore/src/miner/transaction_queue.rs]: Fixed.

**Pull request merged**